src/heroic.rs

```diff
@@ -0,0 +1,252 @@
+                        legendary_installed
+                    );
+                }
+            }
+        }
+    }
+
+    fn detect_gog_roots(&mut self, root: &RootsConfig) {
+        log::trace!(
+            "detect_gog_roots searching for GOG information in {}",
+            root.path.interpret()
+        );
+
+        // use gog_store/library.json to build map .app_name -> .title
+        let mut game_titles = std::collections::HashMap::<String, String>::new();
+        let gog_library = serde_json::from_str::<GogLibrary>(
+            &std::fs::read_to_string(format!("{}/gog_store/library.json", root.path.interpret())).unwrap_or_default(),
+        );
+        gog_library.unwrap().games.iter().for_each(|game| {
+            game_titles.insert(game.app_name.clone(), game.title.clone());
+        });
+        log::trace!(
+            "detect_gog_roots found {} games in {}",
+            game_titles.len(),
+            format!("{}/gog_store/library.json", root.path.interpret())
+        );
+
+        // iterate over all games found in HEROCONFIGDIR/gog_store/installed.json and call find_game_root
+        let content = std::fs::read_to_string(format!("{}/gog_store/installed.json", root.path.interpret()));
+        if let Ok(installed_games) = serde_json::from_str::<HeroicInstalled>(&content.unwrap_or_default()) {
+            installed_games.installed.iter().for_each(|game| {
+                let game_title = game_titles.get(&game.game_id).unwrap();
+                if let Some(sp) = self.find_game_root(root.path.interpret(), game_title, &game.platform, &game.game_id)
+                {
+                    self.memorize_game_root(game_title, &sp);
+                }
+            })
+        }
+    }
+
+    fn memorize_game_root(&mut self, title: &String, path: &StrictPath) {
+        let normalized = normalize_title(title);
+        if let Some(official) = self.normalized_to_official.get(&normalized) {
+            log::trace!(
+                "memorize_game_root memorizing path {} for {}",
+                path.interpret(),
+                official
+            );
+            self.games.insert(official.clone(), path.clone());
+        } else {
+            log::warn!(
+                "memorize_game_root did not find {} in manifest, no backup/restore will be done!",
+                title
+            );
+            log::trace!("memorize_game_root memorizing path {} for {}", path.interpret(), title);
+            self.games.insert(title.clone(), path.clone());
+        }
+    }
+
+    fn find_game_root(
+        &self,
+        heroic_path: String,
+        game_name: &String,
+        platform: &String,
+        game_id: &String,
+    ) -> Option<StrictPath> {
+        match platform.as_str() {
+            "windows" => {
+                // no struct for type safety used here since GamesConfig use the game id as a key name
+                let v: serde_json::Value = serde_json::from_str(
```

Comment:
I researched this and it would be the correct approach, if only the json would be "well" formed.  `GameConfig/*.json` look like:

```json
{
  "1207663333": {
    "autoInstallDxvk": false,
    "autoInstallVkd3d": false,
    "preferSystemLibs": false,
    "nvidiaPrime": false,
    "enviromentOptions": [],
    "wrapperOptions": [
      {
        "exe": "/home/MYUSER/Projekte/workspace/ludusavi-cloud/ludusavi-cloud.sh"
        "args": ""
      }
    ],
    "showFps": false,
    "useGameMode": false,
    "language": "",
    "wineVersion": {
      "bin": "/usr/bin/wine",
      "name": "Wine Default - wine-ge-7.0 (Staging)",
      "type": "wine",
      "wineserver": "/usr/bin/wineserver",
      "wineboot": "/usr/bin/wineboot"
    },
    "winePrefix": "/home/MYUSER/Games/Heroic/Prefixes/Avernum"
  },
  "version": "v0",
  "explicit": true
}
```

So the first maps a dynamic key to a GamesConfig, the others map to string.  I could make the wrapper a `struct GamesConfigWrapper(HashMap<String, String>)` and then parse the value associated with of the `app_name` to a `GamesConfig`.  That would be a compromise.

Reply:
Ah, yeah, that's inconvenient. If you put the value type as `String`, I'd expect it to reject/discard the object and boolean values.

For cases like this, you can use an enum to preserve type-safety:

```rust
#[derive(serde::Deserialize)]
struct GamesConfigWrapper(HashMap<String, GamesConfig>);

#[derive(serde::Deserialize)]
enum GamesConfig {
    Standard {
        // winePrefix, etc
    }
    #[serde(other)]
    Other
}
```

If we ever needed to check the `version` field, for example, we could add another variant like `Text(String)`.

Reply:
Indeed it rejected it, kindly telling me that `find_prefix error invalid type: map, expected a string at line 2 column 16, ignoring`.

The solution you recommend does not (yet?) work either.  Serde untagged does not like Other and fails compilation with `error: #[serde(other)] cannot appear on untagged enum`.

So I tried this:

```rust
#[derive(serde::Deserialize, Debug)]
struct GamesConfigWrapper(HashMap<String, GamesConfig>);

#[derive(serde::Deserialize, Debug)]
#[serde(untagged)]
enum GamesConfig {
    Config {
        #[serde(rename = "winePrefix")]
        wine_prefix: String,
        #[serde(rename = "wineVersion")]
        wine_version: GamesConfigWine,
    },
    Version {
        version: String,
    },
    Explicit {
        explicit: bool,
    },
}
```
It compiles, but does not like the json yet:  `find_prefix error: 'data did not match any variant of untagged enum GamesConfig at line 26 column 17', ignoring`.

Reply:
Could you try taking my version and just removing the `#[serde(other)]` line?

Reply:
Got it.  A working solution is this:

```rust
#[derive(serde::Deserialize, Debug)]
struct GamesConfigWrapper(HashMap<String, GamesConfig>);

#[derive(serde::Deserialize, Debug)]
#[serde(untagged)]
enum GamesConfig {
    Config {
        #[serde(rename = "winePrefix")]
        wine_prefix: String,
        #[serde(rename = "wineVersion")]
        wine_version: GamesConfigWine,
    },
    Version(String),
    Explicit(bool),
}

#[derive(serde::Deserialize, Debug)]
struct GamesConfigWine {
    #[serde(rename = "type")]
    wine_type: String,
}
```

Reply:
Better solution ignoring everything except the wanted data structure:

```rust
#[derive(serde::Deserialize, Debug)]
struct GamesConfigWrapper(HashMap<String, GamesConfig>);

#[derive(serde::Deserialize, Debug)]
#[serde(untagged)]
enum GamesConfig {
    Config {
        #[serde(rename = "winePrefix")]
        wine_prefix: String,
        #[serde(rename = "wineVersion")]
        wine_version: GamesConfigWine,
    },
    Other(serde::de::IgnoredAny),
}

#[derive(serde::Deserialize, Debug)]
struct GamesConfigWine {
    #[serde(rename = "type")]
    wine_type: String,
}
```